test/BIP324: disconnection scenarios during v2 handshake

stratospher opened the pull request:
Add tests for the following v2 handshake scenarios:
1. Disconnection happens when > `MAX_GARBAGE_LEN` bytes garbage is sent
2. Disconnection happens when incorrect garbage terminator is sent
3. Disconnection happens when garbage bytes are tampered with
4. Disconnection happens when AAD of first encrypted packet after the garbage terminator is not filled
5. bitcoind ignores non-empty version packet and no disconnection happens

All these tests require a modified v2 P2P class (different from `EncryptedP2PState` used in `v2_p2p.py`) to implement our custom handshake behaviour based on different scenarios and have been kept in a single test file (`test/functional/p2p_v2_misbehaving.py`). Shifted the test in `test/functional/p2p_v2_earlykeyresponse.py` which is of the same pattern to this file too.

stratospher committed:
test: Rename early key response test and move random_bitflip to util

Early key response test is a special kind of test which requires
modified v2 handshake functions. More such tests can be added
where v2 handshake functions send incorrect garbage terminator,
excess garbage bytes etc.. Hence, rename p2p_v2_earlykey.py to a
general test file name - p2p_v2_misbehaving.py.

random_bitflip function (used in signature tests prior to this
commit) can be used in p2p_v2_misbehaving test to generate wrong
garbage terminator, wrong garbage bytes etc..
So, move the function to util.

stratospher committed:
test: Introduce test types and modify v2 handshake function accordingly

Prior to this commit, TestEncryptedP2PState would always
send initial_v2_handshake bytes in 2 parts (as required
by early key response test).

For generalising this test and having different v2 handshake
behaviour based on the test type, special behaviours like
sending initial_v2_handshake bytes in 2 parts are executed
only if test_type is set to EARLY_KEY_RESPONSE.

stratospher committed:
test: Support disconnect waiting for add_p2p_connection

Adds a new boolean parameter `expect_success` to the
`add_p2p_connection` method. If set, the node under
test doesn't wait for connection to be established
and is useful for testing scenarios when disconnection
is expected.

Without this parameter, intermittent test failures can happen
if the disconnection happens before wait_until for is_connected
is hit inside `add_p2p_connection`.

Co-Authored-By: Sebastian Falbesoner <sebastian.falbesoner@gmail.com>

stratospher committed:
test: Check that disconnection happens when >4095 garbage bytes is sent

This test type is represented using EXCESS_GARBAGE.

stratospher committed:
test: Check that disconnection happens when wrong garbage terminator is sent

This test type is represented using WRONG_GARBAGE_TERMINATOR.
since the wrong garbage terminator is sent to TestNode, TestNode
will interpret all of the gabage bytes, wrong garbage terminator,
decoy messages and version packet it receives as garbage bytes.

If the length of all these is more than 4095 + 16, it will result
in a missing garbage terminator error. otherwise, it will result
in a V2 handshake timeout error.

Send only MAX_GARBAGE_LEN//2 bytes of garbage data to TestNode
so that the total length received by the TestNode is at max
= (MAX_GARBAGE_LEN//2) + 16 + 10*120 + 20 = 3283 bytes
(which is less than 4095 + 16 bytes) and we get a consistent
V2 handshake timeout error message.

If we do not limit the garbage length sent, we will intermittently
get both missing garbage terminator error and V2 handshake
timeout error based on the garbage length and decoy packets length
which are chosen at random.

stratospher committed:
test: Check that disconnection happens when garbage sent/received are different

This test type is represented using WRONG_GARBAGE.
Here, garbage bytes sent to TestNode are assumed to be tampered with and
do not correspond to the garbage bytes which P2PInterface calculated and
uses.

stratospher committed:
test: Check that non empty version packet is ignored and no disconnection happens

This test type is represented using SEND_NON_EMPTY_VERSION_PACKET.

stratospher committed:
test: Check that disconnection happens when AAD isn't filled

This test type is represented using SEND_NO_AAD. If AAD of the first encrypted packet
sent after the garbage terminator (optional decoy packet/version packet) hasn't been
filled, disconnection happens.

stratospher commented:
@sr-gi, i tried [this manually sending idea](https://github.com/bitcoin/bitcoin/pull/29352#pullrequestreview-1854149947) but i still think intermittent failures are possible there.
- we can't get rid of `can_data_be_received` variable because if we don't use this variable, test would succeed irrespective of whether we send 4 bytes network magic first or 4 bytes from ellswift bytes first and we don't want that.
- so since `data_received()` always happens in `Network thread`  and send of ellswift bytes + setting  `can_data_be_received=True` happens on `MainThread`, in the rare scenario that `data_received()` gets called before setting `can_data_be_received`, an intermittent failure could happen i think.

here's a [branch](https://github.com/stratospher/bitcoin/commit/7e7173467fd5459e4b19ebd48186690f174be95f) where i tweaked the code you shared a bit with a sleep statement for making the test crash and reintroducing `can_data_be_received` variable.

kevkevinpal commented on test/functional/p2p_v2_misbehaving.py:
```diff
@@ -0,0 +1,201 @@
+        elif self.test_type == TestType.WRONG_GARBAGE_TERMINATOR:
+            garbage_len = random.randrange(MAX_GARBAGE_LEN//2)
+        else:
+            garbage_len = random.randrange(MAX_GARBAGE_LEN + 1)
```
doesn't the `+1` make the `garbage_len` greater than `4095` which is what is being covered in `TestType.EXCESS_GARBAGE`

kevkevinpal replied:
never mind randrange doesn't include the value passed

stratospher replied:
yes!

kevkevinpal reviewed (commented):
Concept ACK

added two comments I will continue to test these changes as I am still trying to understand fully what the tests do exactly

stratospher commented on test/functional/p2p_v2_misbehaving.py:
```diff
@@ -0,0 +1,201 @@
+        ]
+        for test_type in TestType:
+            if test_type == TestType.EARLY_KEY_RESPONSE:
+                continue
```
`test_earlykeyresponse` sends handshake bytes etc.. in the `MainThread` and I feel it's better to isolate that behaviour in a different test.

also, if you grep for `def test_` in `test/functional`, you'll see many [examples](https://github.com/bitcoin/bitcoin/blob/3cbc8cbc71d3d6ecfaf41164ce59c24ac94bae99/test/functional/feature_asmap.py#L128) of different tests listed in `run_test()` using the pattern which is currently used.

stratospher commented on test/functional/p2p_v2_misbehaving.py:
```diff
@@ -0,0 +1,201 @@
+        self.extra_args = [["-v2transport=1", "-peertimeout=3"]]
+
+    def run_test(self):
+        self.test_earlykeyresponse()
+        self.test_v2disconnection()
```
yeah! `expected_debug_message` on [L180](https://github.com/bitcoin/bitcoin/blob/ec9005ca4be088dfa2d247bbfe964a9c98e4f29d/test/functional/p2p_v2_misbehaving.py#L180) was written with the assumption that `test_v2disconnection` is run after `test_earlykeyresponse`. so peer=0 happens in `test_earlykeyresponse` and peer=1,2,3,4,5 happens in  `test_v2disconnection`.

sr-gi commented:
> * we can't get rid of `can_data_be_received` variable because if we don't use this variable, test would succeed irrespective of whether we send 4 bytes network magic first or 4 bytes from ellswift bytes first and we don't want that.

I've been playing around sending anything different than the first 4 bytes as the network magic and test fail on my end (which is expected given `data_received` is called with `magic_sent` false on the first-byte mismatch.

Do you have any working example I can try to reproduce?

stratospher commented:
> I've been playing around sending anything different than the first 4 bytes as the network magic and test fail on my end (which is expected given data_received is called with magic_sent false on the first-byte mismatch).

yes! that's expected behaviour and sending anything other than the first 4 bytes as network magic will make the test fail on master.

what i was trying to say was:
1. on master
- replacing `return b"\xfa\xbf\xb5\xda"` with any other 4 bytes will make the test fail and this is expected behaviour
3. using [this patch](https://github.com/bitcoin/bitcoin/pull/29352#pullrequestreview-1854149947) mentioned in #29352
- replacing `return b"\xfa\xbf\xb5\xda"` with any other 4 bytes will make the test pass and this isn't behaviour we want
- this happens because [`can_data_be_received` variable](https://github.com/bitcoin/bitcoin/blob/ab5dfdbec1143f673f4d83acd4e335bb2c51034e/test/functional/p2p_v2_earlykeyresponse.py#L29) is removed in the patch

so i was just trying to say that we can't remove [`can_data_be_received` variable](https://github.com/bitcoin/bitcoin/blob/ab5dfdbec1143f673f4d83acd4e335bb2c51034e/test/functional/p2p_v2_earlykeyresponse.py#L29).

mzumsande reviewed (commented):
Concept ACK

The one thing I'm a bit unsure about is that this duplicates quite a lot of code, especially in `complete_handshake` - the main implementation in `v2_p2p.py` and the overwritten version from this test could run out of sync in the future. But I don't have a good idea how to avoid this yet...

stratospher commented on test/functional/p2p_v2_misbehaving.py:
```diff
@@ -3,87 +3,107 @@
         peer1.send_raw_message(ellswift_and_garbage_data)
-        peer1.wait_for_disconnect(timeout=5)
-        self.log.info('successful disconnection when MITM happens in the key exchange phase')
+        with self.nodes[0].assert_debug_log(['version handshake timeout peer=0']):
```
done.

stratospher commented on test/functional/p2p_v2_misbehaving.py:
```diff
@@ -3,87 +3,107 @@
+    """ Scenarios to be tested:
+
+    1. EARLY_KEY_RESPONSE - The responder needs to wait until one byte is received which does not match the 16 bytes
+    consisting of network magic followed by "version\x00\x00\x00\x00\x00" before sending out it's ellswift + garbage bytes
```
done.

stratospher commented on test/functional/test_framework/test_node.py:
```diff
@@ -694,6 +694,9 @@ def add_p2p_connection(self, p2p_conn, *, wait_for_verack=True, send_version=Tru
         supports_v2_p2p = self.use_v2transport and supports_v2_p2p
         p2p_conn.peer_connect(**kwargs, send_version=send_version, net=self.chain, timeout_factor=self.timeout_factor, supports_v2_p2p=supports_v2_p2p)()
 
+        if wait_for_disconnect:
```
woah more racy thread logic. good catch! i've implemented the `expect_success` parameter suggestion.

stratospher commented:
> The one thing I'm a bit unsure about is that this duplicates quite a lot of code, especially in complete_handshake - the main implementation in v2_p2p.py and the overwritten version from this test could run out of sync in the future. But I don't have a good idea how to avoid this yet...

hmm good point. will need to think about it more. simplest solution would be keeping it all in main implementation in `v2_p2p.py` haha but doesn't make sense to have it there just for this test file :(

sr-gi commented:
I find the approach fairly hard to follow here. Having all the logic in the constructor with if/elses based on the connection type instead of having different constructors/a test for each type of fail feels really error-prone and difficult to make sure if a test is doing what we expect. I think it'd be better to have multiple classes that build from `v2_p2p` and modify what's needed, even if the file gets larger.

Just an example:

In the `WRONG_GARBAGE_TERMINATOR` case, we are modifying the gargabe in two ways. First, we are making it smaller than expected, but also, we are randomly flipping some of its bits. Turns out if you removed both changes, the tests still fails, but I'm not sure why. I would expect this not to be the case, since removing the relevant parts should make the class equal to the expected `v2_p2p`.

stratospher committed:
test: Log when the garbage is actually sent to transport layer

Currently, we log the number of bytes of garbage when it is
generated. The log is a better fit for when the garbage
actually gets sent to the transport layer.

stratospher committed:
test: Make global TRANSPORT_VERSION variable an instance variable

Currently, transport version is a global variable declared as
TRANSPORT_VERSION in v2_p2p.py. Making it an instance variable
would help in sending non empty transport version packets for
testing purposes. It might also help EncryptedP2PState be more
extensible in far future protocol upgrades.

stratospher commented:
> I think it'd be better to have multiple classes that build from v2_p2p and modify what's needed, even if the file gets larger.

that's a great suggestion @sr-gi! it's possible to design those classes in such a way that it avoids code duplication and the file is almost the same size!

> Turns out if you removed both changes, the tests still passes, but I'm not sure why. 

because we don't send a version message to ensure that the disconnection happens in the v2 handshake phase. in your case (different from what we are testing here), we send the correct garbage terminator but because we don't send a version message - disconnection is still expected.

stratospher commented:
I've updated the PR to have different child class implementations for `EncryptedP2PState` based on the disconnection scenario we're testing - `EarlyKeyResponseState`, `ExcessGarbageState` etc...

Also introduced 2 more commits for cleaner code:
- c642b08c - logging when the garbage is sent instead of when garbage is generated (suggestion from https://github.com/bitcoin/bitcoin/pull/24748#discussion_r1466826427 which is useful now that we have multiple child classes for `EncryptedP2PState`)
- 38eb429 - transport version as a class variable instead of a global variable

stratospher commented on test/functional/p2p_v2_misbehaving.py:
```diff
@@ -3,87 +3,107 @@
+            if not self.magic_sent:
+                self.generate_keypair_and_garbage()
+                self.magic_sent = True
+                return b"\xfa\xbf\xb5\xda"
```
done.

sr-gi reviewed (commented):
I like this approach much better, it is way easier to follow and reason about. I left some comments inline.

> because we don't send a version message to ensure that the disconnection happens in the v2 handshake phase. in your case (different from what we are testing here), we send the correct garbage terminator but because we don't send a version message - disconnection is still expected.

What is weird is that the test is expecting a specific log message, which seems to also match in this case. This is still the case by the way. Is there an alternative way of making sure this does not happen?

stratospher committed:
log: Add V2 handshake timeout

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -79,6 +78,7 @@ class EncryptedP2PState:
 
         encrypt/decrypt v2 P2P messages using v2_enc_packet() and v2_receive_packet().
     """
+    transport_version = b''
```
done.

stratospher commented on test/functional/p2p_v2_misbehaving.py:
```diff
@@ -3,87 +3,100 @@
+            # 1. v2 TestNode shouldn't respond back if we send V1_PREFIX and data shouldn't be received on recvbuf.
+            # This state is represented using `can_data_be_received` = False.
+            # 2. v2 TestNode responds back when mismatch from V1_PREFIX happens and data can be received on recvbuf.
+            # This state is represented using `can_data_be_received` = True.
```
> I still think this can be simplified. initiate_v2_handshake is going to be called, asynchronously, when creating this object. You can reduce a lot of the boilerplate, and get rid of magic_sent, as long as you do the last bit manually on the test (or just create another method, like continue_v2_handshake).

yes, that can be done since the last bit is being sent from `MainThread` anyways.

before: 
1. send both 4 bytes network magic (first 4 bytes of ellswift) - `NetworkThread`
2. remaining `ellswift_garbage` bytes - `MainThread`

now:
1. send both 4 bytes network magic (first 4 bytes of ellswift) - `MainThread`
2. remaining `ellswift_garbage` bytes - `MainThread`

realised [my earlier comment](https://github.com/bitcoin/bitcoin/pull/29431#issuecomment-1945395946) is the situation in master too and a sleep in `MainThread` where there's no waiting for some other tasks is unlikely. 😬

I've used [your original suggestion](https://github.com/bitcoin/bitcoin/pull/29352#pullrequestreview-1854149947) and updated the PR to send everything manually!

stratospher commented:
thank you for the helpful reviews @sr-gi!

I've updated the PR:
- `EARLY_KEY_RESPONSE` test doesn't have `magic_sent` now and the `initial_v2_handshake()` function is cleaner
- `EARLY_KEY_RESPONSE` test does sending of 4 bytes ellswift (which match network magic) + remaining ellswift and garbage bytes on `MainThread` - both are done manually now
- Added v2 handshake timeout log in 7d07daa

     
> What is weird is that the test is expecting a specific log message, which seems to also match in this case. This is still the case by the way. Is there an alternative way of making sure this does not happen?

hmm fair point, i've added a log message in `InactivityCheck()` to distinguish v2 handshake timeout case from version handshake timeout.

sr-gi reviewed (commented):
I love it now 🚀 

ACK ae528cb9e29b75d36a8018954f45bcfb886a4557

stratospher commented on test/functional/test_framework/test_node.py:
```diff
@@ -687,14 +687,15 @@ def add_p2p_connection(self, p2p_conn, *, wait_for_verack=True, send_version=Tru
         if supports_v2_p2p is None:
             supports_v2_p2p = self.use_v2transport
 
-
```
`PEP 8: E303 too many blank lines (2)`  was showing up in my code editor. so maybe we can keep it since we're touching the function here.

Mazzika1 replied:
Okay

stratospher commented on test/functional/p2p_v2_misbehaving.py:
```diff
@@ -3,87 +3,79 @@
-        peer1.wait_for_disconnect(timeout=5)
-        self.log.info('successful disconnection when MITM happens in the key exchange phase')
+        peer1.send_raw_message(peer1.v2_state.ellswift_ours[4:] + peer1.v2_state.sent_garbage)
+        peer1.v2_state.can_data_be_received = True
```
agree! done.

stratospher commented on test/functional/p2p_v2_misbehaving.py:
```diff
@@ -32,6 +38,15 @@ def initiate_v2_handshake(self):
+    """Generate > MAX_GARBAGE_LEN garbage bytes"""
+    def generate_keypair_and_garbage(self):
+        ellswift_garbage_bytes = super().generate_keypair_and_garbage()
+        even_more_garbage = random.randbytes(MAX_GARBAGE_LEN)
```
oh looks like 0 garbage bytes being returned can also mess up the functions where we're tampering the garbage bytes (`WRONG_GARBAGE`) and `SEND_NO_AAD` where AAD is the garbage bytes sent. so I've change `generate_keypair_and_garbage()` in `EncryptedP2PState` to return 1 bytes garbage minimum instead.

Mazzika1 replied:
Okay

Mazzika1 replied:
Okay

mzumsande replied:
Removing the possibility of 0 garbage in `EncryptedP2PState` just to make this test nicer is a bit unfortunate in my opinion: If the node had a bug dealing with zero-length garbage, the test framework wouldn't be able to catch that anymore. Can't we change it locally just for the test cases who need it?

stratospher replied:
hmm good point. I've added `garbage_len` as an optional argument in `generate_keypair_and_garbage()` so that some of the test cases which don't support 0 length garbage can avoid it.

stratospher commented on test/functional/p2p_v2_misbehaving.py:
```diff
@@ -3,87 +3,79 @@
         node0 = self.nodes[0]
         self.log.info('Sending first 4 bytes of ellswift which match network magic')
         self.log.info('If a response is received, assertion failure would happen in our custom data_received() function')
         # send happens in `initiate_v2_handshake()` in `connection_made()`
```
done.

mzumsande reviewed (commented):
ACK c9dacd958d7c1e98b08a7083c299d981e4c11193

I'd prefer combining some of the first commits with the respective test commits that actually make use of the change, but that's just my opinion ad maybe a matter of taste.

theStack reviewed (approved):
Code-review ACK c9dacd958d7c1e98b08a7083c299d981e4c11193

Nice approach using a `EncryptedP2PState` subclass for each test scenario, I agree with other reviewers that this looks much cleaner now and is easier to follow.

achow101 commented:
ACK c9dacd958d7c1e98b08a7083c299d981e4c11193

fanquake commented:
Looks like this is causing intermittent CI failures. See #30419.

hebasto commented:
Ported to the CMake-based build system in https://github.com/hebasto/bitcoin/pull/264.
